nexusutils/dataconverter/readers/rii_database/citation_parser.py

```diff
@@ -0,0 +1,113 @@
+import re
+from typing import Any, Dict, Optional
+import requests_cache
+
+
+@dataclass
+class Citation:
+    """A representation for a citation"""
+
+    _raw_str: str = field(repr=False)
+    ref_str: str
+    url: Optional[str]
+    doi: Optional[str]
+    session: Optional[requests_cache.CachedSession] = None
+    bibtex: Optional[str] = None
+
+    @staticmethod
+    def parse_citations(reference: str, download_bibtex: bool = False):
+        """Parses a reference string into its components"""
+        cites = re.split(r"<br\s*\/?\s*>", reference)
+
+        clean_cites = []
+        for cite in cites:
+            clean_cites.append(Citation(cite, download_bibtex))
+        return clean_cites
+
+    def __init__(
+        self, ref_str: str, download_bibtex: bool = False, desc: Optional[str] = None
+    ):
+        self._raw_str = ref_str
+        self.ref_str = re.sub(
+            r"\<\/?[^\<\>]*\>",
+            "",
+            re.search(r"(?:\d\))?\s*(.*)", ref_str).group(1),
+        )
+        self.get_non_doi_url()
+        self.get_doi()
+
+        if desc:
+            self.description = desc
+        else:
+            self.description = (
+                "Literature reference from which this "
+                "dispersion function was extracted."
+            )
+
+        if download_bibtex:
+            self.download_bibtex()
+
+    def download_bibtex(self):
+        """Requests the bibtext entry for a given doi"""
+        if not self.doi:
+            return
+
+        if self.session is None:
+            self.session = requests_cache.CachedSession("rii_bibtex")
+
+        req = self.session.get(
+            f"https://doi.org/{self.doi}",
+            timeout=30,
+            headers={"Accept": "application/x-bibtex"},
+        )
+        if req.status_code == 200:
+            self.bibtex = req.text
+
+    def get_non_doi_url(self):
+        """Extract urls which are not dois"""
+        url = re.search(r"href\=\"([^\"]+)\"", self._raw_str)
+        self.url = url.group(1) if url and "doi.org" not in url.group(1) else None
+
+    def get_doi(self):
+        """Extracts doi from a reference string"""
+        doi_match = re.search(r"\".*doi\.org\/([^\"]+)\"", self._raw_str)
+        self.doi = doi_match.group(1) if doi_match else None
+
+    def write_entries(self, template: Dict[str, Any], path: str):
+        """Write entries to nexusutils template"""
+        reference_identifier = "REFERENCES"
+        name = basename = "reference"
+
+        while f"{path}/{reference_identifier}[{name}]/text" in template:
+            count = locals().get("count", 0) + 1
+            name = f"{basename}{count}"
+
+        base_path = f"{path}/{reference_identifier}[{name}]"
+        if self.url:
+            template[f"{base_path}/url"] = self.url
+        if self.doi:
+            template[f"{base_path}/doi"] = self.doi
+        if self.bibtex:
+            template[f"{base_path}/bibtex"] = self.bibtex
```

Comment:
Do you think we should add this behavior to the template as a default:

```python
__set__(self, instance, value):
   if value:
      self.value = value
```

Reply:
I feel it's better to keep it explicit. Here, it wouldn't even work as this is a "normal" dict with which the actual template of type `Template` gets updated with.

What would happen with `None` values in the template being returned by the read function? Would this result in an error or would the keys just be ignored for non-required fields? Maybe this could be a good default behaviour, it the field is explicitly None (and not None-like, like "") to just drop the entry if it is not required.

Reply:
The __set__ was wrong. I actually was imagining this one here: https://github.com/nomad-coe/nomad-parser-nexus/blob/16614122a281ce5f5d245c90130ede147c988510/nexusutils/dataconverter/template.py#L66

I say we check ```v``` there and if it is ```None``` we don't do anything. So the previously set key by the generation stays as is. No new ```None``` key will be added. I think I also check requirements explicitly in case someone makes a new template and sends that back in the read function.

Reply:
Ah yes this would work. I just realised that it is doing exactly what I wrote above, because updating the template would just use `__setitem__`. So yes, let's do it like this.

Reply:
I added a check. However, it lets the `test_add_entry` fail. I couldn't really see why, because it only adds to the internal dicts, which should not be affected by setitem. @shabihsherjeel any idea?

Reply:
I commited the failing change so you can see it in the ci/cd

Reply:
I'm fixing these. I'll push a commit in a bit.